[storage::huawei::oceanstor::snmp::plugin] add lun discover and check usage

christophe-activiumid opened the pull request:
adding huawei oceanstor "lun" usage and discovery

tested on Huawei storage with dorado v6 and v5 OS on multiple hardware

this will get individual lun disk usage on oceanstor device with block (Lun) allocated

the plugin was initially based on the storage-pool plugin, removing "statuses" and adding a new metric "protection" which correspond to the number of bytes all the snapshots and protection feature on oceanstore uses on a lun
and finally I had to rewrite  everything after the selection, as the filter and philosophy behind a lun is different from the pool


Auto-discover is filtering on an OID "hwInfoLunExposedToInitiator" which tells if the lun is presented (usable) to a host
(all snaphot are Lun not exposed, it makes no sense to discover or list these lun)

I added a --include-non-exposed-luns on the check command, in case someone wants that

if there is a way to filter based on a macro in the discovery rule, I have implemented a "isdiscovered" macro in the discover command, but no idea how to filter that in the discovery rules, if it is filterable in the UI then the line 64 in listluns.pm is not necessary

### **implementation**

- check command : HW-Storage-Huawei-Oceanstor-SNMP-luns

`--plugin=storage::huawei::oceanstor::snmp::plugin --mode=luns --hostname='$HOSTADDRESS$' --snmp-version='$_HOSTSNMPVERSION$' --snmp-community='$_HOSTSNMPCOMMUNITY$' $_HOSTSNMPEXTRAOPTIONS$ --filter-name='$_SERVICEFILTERNAME$' --warning-space-usage='$_SERVICEWARNINGSPACEUSAGE$' --critical-space-usage='$_SERVICECRITICALSPACEUSAGE$' --warning-space-usage-free='$_SERVICEWARNINGSPACEUSAGEFREE$' --critical-space-usage-free='$_SERVICECRITICALSPACEUSAGEFREE$' --warning-space-usage-prct='$_SERVICEWARNINGSPACEUSAGEPRCT$' --critical-space-usage-prct='$_SERVICECRITICALSPACEUSAGEPRCT$' $_SERVICEEXTRAOPTIONS$`

- service Template

Template name : HW-Storage-Huawei-Oceanstor-LUN-SNMP (-custom)
Template Alias : Lun-Usage

- discover command: HW-Storage-Huawei-Oceanstor-SNMP-luns-Discover

`--plugin=storage::huawei::oceanstor::snmp::plugin --mode=list-luns --hostname='$HOSTADDRESS$' --snmp-version='$_HOSTSNMPVERSION$' --snmp-community='$_HOSTSNMPCOMMUNITY$' $_HOSTSNMPEXTRAOPTIONS$ --disco-show`

- macro command HW-Storage-Huawei-Oceanstor-SNMP-luns-Macro

`--plugin=storage::huawei::oceanstor::snmp::plugin --mode=list-luns --hostname=127.0.0.1 --disco-format`


- Discovery rule : HW-Storage-Huawei-Oceanstor-SNMP-Lun-Usage-Name

Service display name : Lun-Usage-$name$
host template : HW-Storage-Huawei-Oceanstor-SNMP-custom
MACRO : FILTERNAME=^$name$$

Sims24 commented:
Hi, 

Thanks for this contribution. 

Would it be possible to have mib files and a walk from this branch please: .1.3.6.1.4.1.34774.4.1.23.4.8.1 

You can send it through email if you prefer. 

Also, a question, if I get it right, you tried to add a status check on the exposed attribute of a lun right? That's why some code is commented? 

Thanks,

christophe-activiumid commented:
I'll send you the snmp walk of a test storage by mail.
For the mib, I can't post it here I think, i'm not allowed to share it

for the status, no. I removed the status check that was done on the StoragePool plugin as there is not really a "health status" on the lun

and for the exposed filter : the OID "hwInfoLunExposedToInitiator" is set to 1 if the LUN is presented to a host, 0 if not
I mainly saw this being set to 0 on every SAN Storage snapshot being present, and on Oceanstor, a snapshot is seen as a lun

so if you have a backup systeme doing daily/hourly snapshots, or replication, or anything using snapshot, you will get dozen more entry in the autodiscover list, your LUN and all the snapshot, and this snaphot are mostly transient if you have a software to rotate them
it's not something you want in an audiscover, so by default I filtered them
(you'll see in the snmpwalk)

if you have non exposed lun on a storage, it's not really something you should have except certain use case, and for that I added the possibility to include thes

after a week of usage, I have only one issue : I should have removed the "free space" metric, it's not really useful on the graphic, but that's more cosmetic than anything else
